internal/cms/signed.go

```diff
@@ -0,0 +1,317 @@
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifySignedAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.3 SignerInfo Type
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignedAttributes(signerInfo *SignerInfo, chains [][]*x509.Certificate) ([]*x509.Certificate, error) {
+	if len(chains) == 0 {
+		return nil, VerificationError{Message: "Failed to verify signed attributes because the certificate chain is empty."}
+	}
+
+	// verify attributes if present
+	if len(signerInfo.SignedAttributes) == 0 {
+		if d.ContentType.Equal(oid.Data) {
+			return nil, nil
+		}
+		// signed attributes MUST be present if the content type of the
+		// EncapsulatedContentInfo value being signed is not id-data.
+		return nil, VerificationError{Message: "missing signed attributes"}
+	}
+
+	var contentType asn1.ObjectIdentifier
+	if err := signerInfo.SignedAttributes.Get(oid.ContentType, &contentType); err != nil {
+		return nil, VerificationError{Message: "invalid content type", Detail: err}
+	}
+	if !d.ContentType.Equal(contentType) {
+		return nil, VerificationError{Message: fmt.Sprintf("mismatch content type: found %q in signer info, and %q in signed data", contentType, d.ContentType)}
+	}
+
+	var expectedDigest []byte
+	if err := signerInfo.SignedAttributes.Get(oid.MessageDigest, &expectedDigest); err != nil {
+		return nil, VerificationError{Message: "invalid message digest", Detail: err}
+	}
+	hash, ok := oid.ToHash(signerInfo.DigestAlgorithm.Algorithm)
+	if !ok {
+		return nil, VerificationError{Message: "unsupported digest algorithm"}
+	}
+	actualDigest, err := hashutil.ComputeHash(hash, d.Content)
+	if err != nil {
+		return nil, VerificationError{Message: "hash failure", Detail: err}
+	}
+	if !bytes.Equal(expectedDigest, actualDigest) {
+		return nil, VerificationError{Message: "mismatch message digest"}
+	}
+
+	// sanity check on signing time
+	var signingTime time.Time
+	if err := signerInfo.SignedAttributes.Get(oid.SigningTime, &signingTime); err != nil {
+		if errors.Is(err, ErrAttributeNotFound) {
+			return chains[0], nil
+		}
+		return nil, VerificationError{Message: "invalid signing time", Detail: err}
+	}
+
+	// verify signing time is within the validity period of all certificates
+	// in the chain. As long as one chain is valid, the signature is valid.
+	for _, chain := range chains {
+		if isSigningTimeValid(chain, signingTime) {
+			return chain, nil
+		}
+	}
+
+	return nil, VerificationError{Message: fmt.Sprintf("signing time, %s, is outside certificate's validity period", signingTime)}
+}
+
+// isSigningTimeValid helpes to check if signingTime is within the validity
+// period of all certificates in the chain
+func isSigningTimeValid(chain []*x509.Certificate, signingTime time.Time) bool {
+	for _, cert := range chain {
+		if signingTime.Before(cert.NotBefore) || signingTime.After(cert.NotAfter) {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
This block of code should be after all methods of `ParsedSignedData`.

Reply:
Updated.